core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -2200,4 +2197,222 @@ public void testDocStats() throws IOException {
         assertEquals(0, docStats.getDeleted());
         assertEquals(numDocs, docStats.getCount());
     }
+
+    public void testDoubleDelivery() throws IOException {
+        final ParsedDocument doc = testParsedDocument("test#1", "1", "test", null, 100, -1, testDocumentWithTextField(), new BytesArray("{}".getBytes(Charset.defaultCharset())), null);
+        Engine.Index operation = new Engine.Index(newUid("test#1"), doc, Versions.MATCH_ANY, VersionType.INTERNAL, Engine.Operation.Origin.PRIMARY, System.nanoTime(), true, false);
+        Engine.Index retry = new Engine.Index(newUid("test#1"), doc, Versions.MATCH_ANY, VersionType.INTERNAL, Engine.Operation.Origin.PRIMARY, System.nanoTime(), true, true);
+        if (randomBoolean()) {
+            engine.index(operation);
+            assertEquals(0, engine.getNumVersionLookups());
+            assertNotNull(operation.getTranslogLocation());
+            engine.index(retry);
+            assertEquals(0, engine.getNumVersionLookups());
+            assertNotNull(retry.getTranslogLocation());
+            assertTrue(retry.getTranslogLocation().compareTo(operation.getTranslogLocation()) > 0);
+        } else {
+            engine.index(retry);
+            assertEquals(1, engine.getNumVersionLookups());
+            assertNotNull(retry.getTranslogLocation());
+            engine.index(operation);
+            assertEquals(1, engine.getNumVersionLookups());
+            assertNotNull(retry.getTranslogLocation());
+            assertTrue(retry.getTranslogLocation().compareTo(operation.getTranslogLocation()) < 0);
+        }
+
+        engine.refresh("test");
+        try (Engine.Searcher searcher = engine.acquireSearcher("test")) {
+            TopDocs topDocs = searcher.searcher().search(new MatchAllDocsQuery(), 10);
+            assertEquals(1, topDocs.totalHits);
+        }
+        operation = new Engine.Index(newUid("test#1"), doc, Versions.MATCH_ANY, VersionType.INTERNAL, Engine.Operation.Origin.PRIMARY, System.nanoTime(), true, false);
```

Comment:
well i am just double checking that this work that's all even if our system doesn't do it